[MRG+2] Fix sparse matrix handling in clustering

amueller opened the pull request:
See #4051.

Needs fixing:
- [x] MeanShift
- [x] DBSCAN
- [x] AffinityPropagation

amueller commented:
@jnothman 
The issue in DBSCAN is here:
```
> /home/andy/checkout/scikit-learn/sklearn/cluster/dbscan_.py(234)fit()
-> self.components_ = X[self.core_sample_indices_].copy()
(Pdb) p self.core_sample_indices_
array([], dtype=int64)
```

jnothman commented:
Huh? Oh, perhaps that's the only part of the implementation I didn't test! :s

amueller commented:
You can never test enough ;)

amueller commented:
fixed the dbscan issue.

jnothman commented on sklearn/cluster/tests/test_dbscan.py:
```diff
@@ -78,6 +77,19 @@ def test_dbscan_sparse():
     assert_array_equal(labels_dense, labels_sparse)
 
 
+def test_dbscan_no_score_samples():
```
*no_core_samples

jnothman replied:
(Maybe that's where the missing s went from a very different patch)

amueller replied:
haha I bet.

ogrisel replied:
This comment has not been addressed.

amueller replied:
Damn. I forgot to push at some point or a rebase messed it up :-/

amueller replied:
done

jnothman commented on sklearn/cluster/dbscan_.py:
```diff
@@ -231,7 +231,11 @@ def fit(self, X, sample_weight=None):
+        if len(self.core_sample_indices_):
+            self.components_ = X[self.core_sample_indices_].copy()
+        else:
+            # no core samples
```
Note that this is to fix a lack in scipy.sparse

jnothman commented:
Except for `check_sparsify_multiclass_classifier` needing a better name, and the special-casing in dbscan needing a comment, this LGTM.

amueller commented:
The ``check_sparsify`` has a better name in the other PR ;) I need to rebase this one, sorry.

amueller commented:
But I can also ditch the other one and we just use this.

amueller commented:
the problem is that I used #4058 elsewhere, too.

jnothman commented:
Ah sorry, I'd forgotten this was on top of #4058

On 12 January 2015 at 05:45, Andreas Mueller <notifications@github.com>
wrote:

>  The check_sparsify has a better name in the other PR ;) I need to rebase
> this one, sorry.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4052#issuecomment-69505558>
> .
>

ogrisel commented:
#4058 has been merged but the diff view of this PR does not reflect it. @amueller can you please rebase this on top of the current master and `push -f` to get github to compute a new diff (and have travis re-run the tests just to be sure?).

amueller commented:
done

ogrisel commented on sklearn/cluster/tests/test_dbscan.py:
```diff
@@ -78,6 +77,19 @@ def test_dbscan_sparse():
+    X[X < .8] = 0
+
+    db = DBSCAN().fit(X)
+    assert_array_equal(db.components_, np.empty((0, X.shape[1])))
```
I assume that all samples are clustered as outliers in that case? It would be great to add an assertion to check that here, first by asserting that `core_sample_indices_` has shape `(0,)` and second that `labels_` is filed with `-1`s.

amueller replied:
done

ogrisel commented:
Apart from this last minor comment, +1 on my side as well.

amueller committed:
TST add test for sparse matrix handling in clustering

Make sparsity check check everything.

don't test everything. That would be nice but is out of scope :-/

catch special case of no core samples in DBSCAN

add nonregression test for sparse dbscan with no core samples.

amueller commented:
will merge if travis agrees with me.
